backend/src/main/scala/com/softwaremill/adopttapir/starter/content/ContentService.scala

```diff
@@ -0,0 +1,51 @@
+package com.softwaremill.adopttapir.starter.content
+
+import com.softwaremill.adopttapir.logging.FLogging
+import com.softwaremill.adopttapir.starter.StarterDetails
+import com.softwaremill.adopttapir.template.ProjectGenerator
+
+class ContentService(projectGenerator: ProjectGenerator) extends FLogging {
+
+  def generateContentTree(starterDetails: StarterDetails): Node = {
+    val projectName = starterDetails.projectName
+    val projectAsFilesWithPaths = projectGenerator.generate(starterDetails)
+    val projectAsDirTrees = projectAsFilesWithPaths.map(gf => {
+      val pathsWithRootDir = (projectName + "/" + gf.relativePath).split('/').toList
+      val content = gf.content
+      createTree(pathsWithRootDir, content)
+    }).map(_.asInstanceOf[Directory])
+    val projectAsTree = projectAsDirTrees.reduce(merge)
+    projectAsTree
+  }
+
+  private def createTree(pathNames: List[String], content: String): Node = {
+    pathNames match {
+      case Nil => throw new IllegalStateException("pathNames list cannot be empty!")
+      case last :: Nil => File(last, content)
+      case head :: tail => Directory(head, List(createTree(tail, content)))
+    }
+  }
+
+  private def merge(n1: Directory, n2: Directory): Directory = {
+    val files = n1.content.filter(_.isInstanceOf[File]) ++ n2.content.filter(_.isInstanceOf[File])
+    val n1Dirs = n1.content.filter(_.isInstanceOf[Directory]).map(_.asInstanceOf[Directory])
+    val n1DirsNames = n1Dirs.map(_.name)
+    val n2Dirs = n2.content.filter(_.isInstanceOf[Directory]).map(_.asInstanceOf[Directory])
+    val n2DirsNames = n2Dirs.map(_.name)
+
+    val n1UniqueDirs = n1Dirs.filter(n1d => !n2DirsNames.contains(n1d.name))
```

Comment:
not sure what happens here but maybe it would be much simpler to create a `Set` of all directory names?

Reply:
Please look at the updated `merge` logic in `ContentMerger,` I hope it looks clearer.